pyiceberg/table/update/snapshot.py

```diff
@@ -279,6 +296,30 @@ def _commit(self) -> UpdatesAndRequirements:
             (AssertRefSnapshotId(snapshot_id=self._transaction.table_metadata.current_snapshot_id, ref="main"),),
         )
 
+    def _validate(self, starting_snapshot: Snapshot, current_snapshot: Snapshot) -> None:
+        # Define allowed operations for each type of operation
+        allowed_operations = {
+            Operation.APPEND: {Operation.APPEND, Operation.REPLACE, Operation.OVERWRITE, Operation.DELETE},
+            Operation.REPLACE: {Operation.APPEND},
```

Comment:
```suggestion
            Operation.REPLACE: {},
```

I think the spec may need a re-review because I think it's inaccurate to say that we only need to verify that the files we are trying to delete are still available when we are executing a `REPLACE` or `DELETE` operation.

In Spark, we also validate whether there's been a conflicting appends when we use `SERIALIZABLE` isolation level:

https://github.com/apache/iceberg/blob/9fc49e187069c7ec2493ac0abf20f73175b3df89/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java#L368-L374

I think it would be helpful to introduce all three types of isolation levels `NONE`, `SERIALIZABLE` and `SNAPSHOT`, and verify if conflicting appends or deletes have been introduced in the underlying partitions to be aligned with the implementation in Spark

Reply:
Thanks @sungwy for jumping in here, and creating the issues 🙌 

And indeed, based on if we do snapshot or serializable isolation, we should allow for new data (or not)